compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ExpressionCodegen.kt

```diff
@@ -265,23 +266,47 @@ class ExpressionCodegen(
         visitStatementContainer(body, data).discard()
 
     override fun visitContainerExpression(expression: IrContainerExpression, data: BlockInfo) =
-        visitStatementContainer(expression, data).coerce(expression.asmType)
+        visitStatementContainer(expression, data).coerce(expression.type)
+
+    private val IrFunctionAccessExpression.isErasedNewCall: Boolean
+        get() = symbol.owner.isPrimaryInlineClassConstructor && irFunction.origin != IrDeclarationOrigin.SYNTHETIC_INLINE_CLASS_MEMBER
+
+    private fun generateErasedNewCall(expression: IrFunctionAccessExpression, data: BlockInfo): PromisedValue {
+        val function = expression.symbol.owner
+        val argument = expression.getValueArgument(0) ?: function.valueParameters[0].defaultValue!!
+        val argumentType = function.valueParameters[0].type
+        val value = argument.accept(this, data).coerce(argumentType).materialized
+        value.irType = function.returnType
+        return value
+    }
 
     override fun visitFunctionAccess(expression: IrFunctionAccessExpression, data: BlockInfo): PromisedValue {
         expression.markLineNumber(startOffset = true)
         classCodegen.context.irIntrinsics.getIntrinsic(expression.symbol)
-            ?.invoke(expression, this, data)?.let { return it.coerce(expression.asmType) }
+            ?.invoke(expression, this, data)?.let { return it.coerce(expression.type) }
+
+        if (expression.isErasedNewCall)
+            return generateErasedNewCall(expression, data)
+
+        if (expression.symbol.owner.isInlineClassFieldGetter) {
+            val argument = expression.dispatchReceiver!!
+            val type = expression.symbol.owner.dispatchReceiverParameter!!.type
+            val value = argument.accept(this, data).coerce(type).materialized
+            value.irType = expression.type
```

Comment:
Doesn't `expression.type` map to the argument's type? Would `.coerce(expression.type)` do something different?

Reply:
Yes it would - there is an implicit coercion between irTypes going on here, which does not change the underlying asm types. I changed the code to do this coercion in an intrinsic instead, which should make it clearer.